70_FLIPFluids/app_resources/compute/prepareCellUpdate.comp.hlsl

```diff
@@ -0,0 +1,58 @@
+#include "../common.hlsl"
+#include "../gridUtils.hlsl"
+#include "../descriptor_bindings.hlsl"
+
+struct SPushConstants
+{
+    uint numElements;
+    uint pad;
+};
+
+[[vk::push_constant]] SPushConstants pc;
+
+[[vk::binding(b_pcuGridData, s_pcu)]]
+cbuffer GridData
+{
+    SGridData gridData;
+};
+
+[[vk::binding(b_pcuPInBuffer, s_pcu)]] RWStructuredBuffer<Particle> particleInBuffer;
+[[vk::binding(b_pcuPOutBuffer, s_pcu)]] RWStructuredBuffer<Particle> particleOutBuffer;
+
+[[vk::binding(b_pcuPairBuffer, s_pcu)]] RWStructuredBuffer<uint2> particleCellPairBuffer;
+[[vk::binding(b_pcuGridIDBuffer, s_pcu)]] RWStructuredBuffer<uint2> gridParticleIDBuffer;
+
+[numthreads(WorkgroupSize, 1, 1)]
+void makeParticleCellPairs(uint32_t3 ID : SV_DispatchThreadID)
+{
+    uint p_id = ID.x;
+    uint c_id = worldPosToFlatIdx(particleInBuffer[p_id].position.xyz, gridData);
+
+    particleCellPairBuffer[p_id] = uint2(c_id, p_id);
+}
+
+[numthreads(WorkgroupSize, 1, 1)]
+void setGridParticleID(uint32_t3 ID : SV_DispatchThreadID)
+{
+    uint currp_id = ID.x;
+
+    uint prevp_id = currp_id == 0 ? pc.numElements - 1 : currp_id - 1;
+    uint nextp_id = currp_id == pc.numElements - 1 ? 0 : currp_id + 1;
+    uint currc_id = particleCellPairBuffer[currp_id].x;
+    uint prevc_id = particleCellPairBuffer[prevp_id].x;
+    uint nextc_id = particleCellPairBuffer[nextp_id].x;
+
+    if (currc_id != prevc_id)
+        gridParticleIDBuffer[currc_id].x = currp_id;
+    if (currc_id != nextc_id)
+        gridParticleIDBuffer[currc_id].y = currp_id + 1;
+}
```

Comment:
I can't guess what this does, can you explain it to me?

Reply:
There's a muuuuch better way to do this but it also requires you ditch the radix sort and write a Global Counting sort yourself.

Create a 3D image (with mip-maps!) with format R32_UINT, then following dispatches:
1. clear image to 0s (can also be done with a transfer `vkCmdFill` command)
2. Atomic Increment a texel for each particle inside it (the image is your histogram)
3. Do an _exclusive global_ prefix sum over the entire grid. This can be done by mapping* each 8x8x8 (or 16x16x16 if you want to play the multiple-item per thread game) region to a workgroup, doing a workgroup reduction, storing that in the mipmap level thats +3 (or +4 in case of 16x) levels up, repeat on the mip you've just written till you reach the penultimate mip level (anything with resolution <= 8x8x8 or whatever you chose).
4. Then start doing **exclusive** prefix sums and propagating them down (take high level result, add to item 0 of current level before summing) until you're back at the original mip level 0
5. Now do step (2) again but use the atomic return value, thats your particle output location in the sorted array

Now the left over 3D image, contains an INCLUSIVE (because we took exclusive and added the non-cumulative histogram again) Histogram of your particles. Therefore each cell value represents the `end` (one past the last) particle offset within the sorted range within it, to get the `begin` you need to look at the "previous" (address wise* during prefix sum) cell. Its the same as your `gridParticleIDBuffer` but 50% smaller.

* morton code/ Z-order curve addressing is the sanest for this (requires a PoT and Cuboid grid) otherwise you get lots of headache computing the previous address.

There's also a variant where you the last prefix sum you perform in (3) is **INCLUSIVE**, and you Atomic `--` in step (5) then output the particle to the atomic return value -1 (post decrement value). That results in the left-over image being an EXCLUSIVE scan which may be easier to work with because that gives you the `begin` particle offset and you can iterate over them until you hit a particle whose position derived gridID no longer matches instead of trying to look for a prefix-sum-addressing-next cell to find your `end`. Although to play devils advocate iterating from `end-1` backwards on an INCLUSIVE scan image until a similar condition is reached also works.

Then you don't need a particle ID/pair buffer.

P.S. The Global Reduction could actually be done in a single dispatch if you are able to detect which workgroup is the last to provide an input for the next stage, and take over the computation of the next stage - can be done via an extra small scratch buffer of atomics. See @kpentaris's work.

P.S. Global Prefix sum can be done in 2 dispatches, but thats outside the scope of this task.